implementations/rust/ockam/ockam_macros/src/node_test_attribute.rs

```diff
@@ -70,35 +70,66 @@ fn output(mut cont: Container) -> TokenStream {
+                        let _ = std::panic::AssertUnwindSafe(async { ctx.stop().await.unwrap(); })
+                            .catch_unwind()
+                            .await;
+                        Err(err)
+                    }
+                }
+            }
 
             let (mut #ctx_ident, mut executor) = NodeBuilder::without_access_control().build();
             executor
                 .execute(async move {
-                    match timeout(Duration::from_millis(#timeout_ms), #test_fn_ident(&mut #ctx_ident)).await {
-                        Ok(r) => match r {
-                            Err(err) => {
-                                #ctx_stop_stmt
-                                Err(err)
-                            },
-                            Ok(_) => Ok(())
-                        },
-                        Err(_) => {
-                            #ctx_stop_stmt
-                            panic!("Test timeout")
+                    let catch_res = std::panic::AssertUnwindSafe(async {
+
+                        let test_result
+                                = timeout(Duration::from_millis(#timeout_ms), #test_fn_ident(&mut #ctx_ident))
+                            .await
+                            .map_err(|_| ockam_core::Error::new(ockam_core::errcode::Origin::Node,
+                                                    ockam_core::errcode::Kind::Timeout, "Test timeout") )
+                            // unwrap timeout result to get inner test result
+                            .unwrap();
+                        // stop ctx and pipe result back
+                        stop_map_err(&mut #ctx_ident, test_result).await
+                    })
+                    .catch_unwind()
+                    .await;
+                    // unwrap result from catch_unwind
+                    match catch_res {
+                        // not panicked, get test result
+                        Ok(test_result) => test_result,
+
+                        Err(test_panic_cause) => {
+                            // ?? ignoring possible error returned by ctx.stop() in sake of test_panic_cause
+                            let _ = #ctx_stop_stmt;
+                            // NOTICE: still panicking in an async context after catch_unwinding
+                            // could cause hanging ?
+                            // would be better to return Err ?, like so:
+                            // ockam_core::Result::<()>::Err("Test panicked".into())
+                            // however no conversion from &str to Err is there
+                            panic!("Test panicked, caused by {:?}", test_panic_cause);
                         }
                     }
                 })
                 .expect("Executor should not fail")
-                .expect("Test function should not fail");
+                .expect("Test function should not fail")
         }
     }).expect("Parsing failure");
```

Comment:
I think you can simplify this a lot. Check this out and let me know what you think!
```rust
cont.original_fn.block = parse2(quote! {
    {
        use core::panic::AssertUnwindSafe;
        use core::time::Duration;
        use #ockam_crate::{NodeBuilder, compat::tokio::time::timeout, compat::futures::FutureExt};

        let (mut #ctx_ident, mut executor) = NodeBuilder::without_access_control().build();
        executor
            .execute(async move {
                if let Err(err) =
                    AssertUnwindSafe(timeout(Duration::from_millis(#timeout_ms), #test_fn_ident(&mut #ctx_ident)))
                        .catch_unwind()
                        .await
                {
                    #ctx_stop_stmt
                    panic!("Test returned error {:?}", err);
                }
            })
            .map_err(|err| panic!("Test returned error {:?}", err))
            .unwrap();
    }
}).expect("Parsing failure");
```

Reply:
Hi @adrianbenavides, 
as for your code suggestion,
the problem here is that we need to deal with both Results::Errs and inner panics. So whenever the test panics inside, catch_rewind converts it into an Error. So your code line if let Err(err)... actually checks if the test panicked, but if the test returns an error, catch_rewind returns Ok. So we need to unwrap it and check the inner Result.
The second point is that you suggest to re-panic upon catch_rewind error thus getting back to the same situation when a code running in async context panics again.
I tried your code, it hangs whenever the inner test wants to return an Error.
As for commiting messages, it is still terra incognita for me, but I'm trying hard :)

Reply:
> it is still terra incognita for me, but I'm trying hard :)

😀 don't worry too much about it, we can correct it for you just before land the PR.

Reply:
Ohh I see, I didn't understand how `catch_unwind` worked then! I'm going to try something else... now I'm curious :laughing:

Reply:
Yes, there are a couple of corner cases:
1) no implicit conversion from timeout::Result to ockam::Result, I needed to map the error
2) it would be nice to get rid of explicit panicking in that snippet while still in async block. My snippet still has that flaw, because I could find which kind of ockam::Error to return in this case. I left a long comment in that block
3) yesterday you said that we must assume that ctx.stop may panic too, so the code deals with it
4) not directly related to this issue, rather to #3105, there are scenarios when the code hangs if ctx.stop called twice. When tha issue is resolved, probably this code also may be simplified.

Reply:
I've updated the snippet from my previous comment https://github.com/build-trust/ockam/pull/3093#discussion_r937578186 with basically your solution but without that extra function. I think this will fix most of the cases where we have those flaky tests hanging in CI, so we can leave it here and see how it goes.

Reply:
Hi @adrianbenavides !
I ran your snippet, it compiles with warnings of unused Result, I'll see tomorrow how to advance with this, it's too late now :)

Reply:
@Retamogordo you did great with the process 👏 👏 
I squashed your commits, rebased and amended the commit message.
When @adrianbenavides @SanjoDeundiak are back online in a some hours, we can merge after their final okay 👍